flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java

```diff
@@ -0,0 +1,316 @@
+
+	@GuardedBy("lock")
+	private JobGraph jobGraph;
+
+	// ------------------------------------------------------------------------
+	//  Program entry point
+	// ------------------------------------------------------------------------
+
+	/**
+	 * The entry point for the YARN application master.
+	 *
+	 * @param args The command line arguments.
+	 */
+	public static void main(String[] args) {
+		EnvironmentInformation.logEnvironmentInfo(LOG, "YARN ApplicationMaster runner", args);
+		SignalHandler.register(LOG);
+		JvmShutdownSafeguard.installAsShutdownHook(LOG);
+
+		// run and exit with the proper return code
+		int returnCode = new YarnFlinkApplicationMasterRunner().run(args);
+		System.exit(returnCode);
+	}
+
+	@Override
+	protected int runApplicationMaster(Configuration config) {
+
+		try {
+			// ---- (1) create common services
+
+			// try to start the rpc service
+			// using the port range definition from the config.
+			final String amPortRange = config.getString(
+					ConfigConstants.YARN_APPLICATION_MASTER_PORT,
+					ConfigConstants.DEFAULT_YARN_JOB_MANAGER_PORT);
+
+			synchronized (lock) {
+				haServices = HighAvailabilityServicesUtils.createAvailableOrEmbeddedServices(config);
+				metricRegistry = new MetricRegistry(MetricRegistryConfiguration.fromConfiguration(config));
+				commonRpcService = createRpcService(config, appMasterHostname, amPortRange);
+
+				// ---- (2) init resource manager -------
+				resourceManager = createResourceManager(config);
+
+				// ---- (3) init job master parameters
+				jobManagerRunner = createJobManagerRunner(config);
+
+				// ---- (4) start the resource manager  and job manager runner:
+				resourceManager.start();
+				LOG.debug("YARN Flink Resource Manager started");
+
+				jobManagerRunner.start();
+				LOG.debug("Job Manager Runner started");
+
+				// ---- (5) start the web monitor
+				// TODO: add web monitor
+			}
+
+			// wait for resource manager to finish
+			resourceManager.getTerminationFuture().get();
+			// everything started, we can wait until all is done or the process is killed
+			LOG.info("YARN Application Master finished");
+		}
+		catch (Throwable t) {
+			// make sure that everything whatever ends up in the log
+			LOG.error("YARN Application Master initialization failed", t);
+			shutdown(ApplicationStatus.FAILED, t.getMessage());
+			return INIT_ERROR_EXIT_CODE;
+		}
+
+		return 0;
+	}
+
+	// ------------------------------------------------------------------------
+	//  Utilities
+	// ------------------------------------------------------------------------
+
+	protected RpcService createRpcService(
+			Configuration configuration,
+			String bindAddress,
+			String portRange) throws Exception{
+		ActorSystem actorSystem = BootstrapTools.startActorSystem(configuration, bindAddress, portRange, LOG);
+		FiniteDuration duration = AkkaUtils.getTimeout(configuration);
+		return new AkkaRpcService(actorSystem, Time.of(duration.length(), duration.unit()));
```

Comment:
Maybe you could extend `RcpServiceUtils.createRpcService` to also support port ranges.